File: backend/routers/game/polling.py
"""
Polling routes.

Endpoint for polling game updates (messages, state changes).
"""

import json
import logging
from typing import Optional

import crud
import schemas
from core.dependencies import (
    RequestIdentity,
    get_agent_manager,
    get_request_identity,
)
from domain.entities.agent import is_action_manager
from domain.services.access_control import AccessControl
from domain.services.localization import Localization
from domain.value_objects.enums import Language, MessageRole, WorldPhase
from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException
from infrastructure.database.connection import async_session_maker, get_db
from orchestration import get_trpg_orchestrator
from sdk import AgentManager
from services.persistence_manager import PersistenceManager
from services.player_service import PlayerService
from services.transient_state_service import TransientStateService
from services.world_service import WorldService
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger("GameRouter.Polling")

router = APIRouter()


@router.get("/{world_id}/poll")
async def poll_updates(
    world_id: int,
    background_tasks: BackgroundTasks,
    since_message_id: Optional[int] = None,
    poll_onboarding: bool = False,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
    agent_manager: AgentManager = Depends(get_agent_manager),
):
    """
    Poll for updates (new messages, state changes).

    This is the main polling endpoint for the frontend.
    Returns new messages since the given ID and current game state.

    Args:
        poll_onboarding: If True, always poll from onboarding room (used when
                         user is still on onboarding page after phase changes to active)

    The filesystem is the source of truth for world phase.
    If the filesystem phase differs from the database, we sync it.
    """
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    # Check filesystem config (source of truth) and sync if needed
    fs_config = WorldService.load_world_config(world.name)
    current_phase = world.phase
    if fs_config:
        # Check which fields need syncing
        updates = {}
        if fs_config.phase != world.phase:
            logger.info(f"Phase mismatch for world {world.name}: DB={world.phase}, FS={fs_config.phase}. Syncing.")
            updates["phase"] = fs_config.phase
            current_phase = fs_config.phase
        if fs_config.user_name and fs_config.user_name != world.user_name:
            logger.info(
                f"user_name mismatch for world {world.name}: DB={world.user_name}, FS={fs_config.user_name}. Syncing."
            )
            updates["user_name"] = fs_config.user_name
        if fs_config.genre and fs_config.genre != world.genre:
            logger.info(f"genre mismatch for world {world.name}: DB={world.genre}, FS={fs_config.genre}. Syncing.")
            updates["genre"] = fs_config.genre
        if fs_config.theme and fs_config.theme != world.theme:
            logger.info(f"theme mismatch for world {world.name}: DB={world.theme}, FS={fs_config.theme}. Syncing.")
            updates["theme"] = fs_config.theme

        # Apply updates if any
        if updates:
            await crud.update_world(db, world_id, schemas.WorldUpdate(**updates))
            # Refresh world object to get updated values
            world = await crud.get_world(db, world_id)

    player_state = await crud.get_player_state(db, world_id)

    # Determine which room to poll based on phase and poll_onboarding flag
    target_room_id = None
    location = None

    # If poll_onboarding is True OR phase is onboarding, use onboarding room
    # This allows the frontend to stay on onboarding page after phase changes
    if (poll_onboarding or current_phase == WorldPhase.ONBOARDING) and world.onboarding_room_id:
        # Poll from the onboarding room
        target_room_id = world.onboarding_room_id
    elif current_phase == WorldPhase.ACTIVE and not poll_onboarding:
        # Check if we need to sync from filesystem (e.g., after onboarding completed via MCP tools)
        if player_state and not player_state.current_location_id:
            # Database doesn't have current location - try to sync from filesystem
            fs_state = PlayerService.load_player_state(world.name)
            if fs_state and fs_state.current_location:
                logger.info(f"Syncing player state from filesystem for world '{world.name}'")
                pm = PersistenceManager(db, world_id, world.name)
                await pm.sync_player_state_from_filesystem()
                # Refresh player_state after sync
                player_state = await crud.get_player_state(db, world_id)

                # Send arrival message to the starting location's room
                if player_state and player_state.current_location_id:
                    arrival_location = await crud.get_location(db, player_state.current_location_id)
                    if arrival_location and arrival_location.room_id:
                        # Get user_name and language from world
                        location_name = arrival_location.display_name or arrival_location.name
                        # Use Korean default if world language is Korean
                        default_name = "여행자" if world.language == Language.KOREAN else "The traveler"
                        user_name = world.user_name if world.user_name else default_name

                        arrival_message = schemas.MessageCreate(
                            content=Localization.get_arrival_message(user_name, location_name, world.language),
                            role=MessageRole.USER,
                            participant_type="system",
                            participant_name="System",
                        )
                        await crud.create_message(
                            db, arrival_location.room_id, arrival_message, update_room_activity=True
                        )
                        logger.info(f"Sent arrival message for '{user_name}' at '{location_name}'")

                        # Trigger initial scene generation in background
                        target_room_for_scene = arrival_location.room_id
                        arrival_content = Localization.get_arrival_message(user_name, location_name, world.language)

                        async def trigger_initial_scene():
                            async with async_session_maker() as session:
                                trpg_orchestrator = get_trpg_orchestrator()
                                task_world = await crud.get_world(session, world_id)
                                if task_world:
                                    await trpg_orchestrator.handle_player_action(
                                        db=session,
                                        room_id=target_room_for_scene,
                                        action_text=arrival_content,
                                        agent_manager=agent_manager,
                                        world=task_world,
                                    )

                        background_tasks.add_task(trigger_initial_scene)
                        logger.info("Polling: Triggered initial scene generation after phase transition")

        # Now try to get the current location's room
        if player_state and player_state.current_location_id:
            location = await crud.get_location(db, player_state.current_location_id)
            if location and location.room_id:
                target_room_id = location.room_id

    if not target_room_id:
        return {"messages": [], "state": None}

    # Check if player is in chat mode
    is_chat_mode = player_state.is_chat_mode if player_state else False

    # Get new messages (use get_messages_since for efficiency if we have since_message_id)
    # In game mode (not chat mode), exclude chat session messages to show only game conversations
    if since_message_id:
        messages = await crud.get_messages_since(db, target_room_id, since_id=since_message_id, limit=50)
        # Filter out chat mode messages if not in chat mode
        if not is_chat_mode and not poll_onboarding:
            messages = [m for m in messages if m.chat_session_id is None]
    else:
        if is_chat_mode or poll_onboarding:
            messages = await crud.get_messages(db, target_room_id)
        else:
            # In game mode, exclude chat session messages
            messages = await crud.get_messages_excluding_chat(db, target_room_id)

    # Filter out system messages (e.g., "Start onboarding..." trigger messages)
    visible_messages = [m for m in messages if m.participant_type != "system"]

    # Load game_time from filesystem (source of truth)
    fs_state = PlayerService.load_player_state(world.name)
    game_time = None
    if fs_state and fs_state.game_time:
        game_time = {
            "hour": fs_state.game_time.get("hour", 8),
            "minute": fs_state.game_time.get("minute", 0),
            "day": fs_state.game_time.get("day", 1),
        }

    # Build response
    response = {
        "messages": [
            {
                "id": m.id,
                "content": m.content,
                "role": m.role,
                "agent_id": m.agent_id,
                "agent_name": m.agent.name if m.agent else None,
                "agent_profile_pic": m.agent.profile_pic if m.agent else None,
                "thinking": m.thinking,
                "timestamp": m.timestamp.isoformat() if m.timestamp else None,
                "image_data": m.image_data,
                "image_media_type": m.image_media_type,
                "game_time_snapshot": json.loads(m.game_time_snapshot) if m.game_time_snapshot else None,
            }
            for m in visible_messages
        ],
        "state": {
            "stats": json.loads(player_state.stats) if player_state and player_state.stats else {},
            "inventory_count": len(json.loads(player_state.inventory))
            if player_state and player_state.inventory
            else 0,
            "turn_count": player_state.turn_count if player_state else 0,
            "phase": current_phase,  # Use synced phase from filesystem
            "pending_phase": fs_config.pending_phase if fs_config else None,  # For "Enter World" button
            "is_chat_mode": player_state.is_chat_mode if player_state else False,  # Chat mode state
            # Resume message ID: when exiting chat mode, frontend should use this as lastMessageId
            # to avoid re-fetching old narration from before chat mode
            "chat_mode_start_message_id": player_state.chat_mode_start_message_id if player_state else None,
            "game_time": game_time,  # In-game time from filesystem
        },
    }

    # Add location info if we have one (not during onboarding)
    if location:
        response["location"] = {
            "id": location.id,
            "name": location.display_name or location.name,
        }

    # Always include suggestions in poll response to avoid race conditions
    # (suggestions may be saved after narration message but before next poll)
    response["suggestions"] = TransientStateService.load_suggestions(world.name)

    return response


@router.get("/{world_id}/chatting-agents")
async def get_chatting_agents(
    world_id: int,
    poll_onboarding: bool = False,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
    agent_manager: AgentManager = Depends(get_agent_manager),
):
    """
    Get list of agents currently generating responses in a world.
    Used by frontend to display thinking/chatting indicators.

    Args:
        poll_onboarding: If True, check onboarding room instead of location room
    """
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    # Determine which room to check based on phase and poll_onboarding flag
    target_room_id = None
    if (poll_onboarding or world.phase == WorldPhase.ONBOARDING) and world.onboarding_room_id:
        target_room_id = world.onboarding_room_id
    elif not poll_onboarding:
        player_state = await crud.get_player_state(db, world_id)
        if player_state and player_state.current_location_id:
            location = await crud.get_location(db, player_state.current_location_id)
            if location and location.room_id:
                target_room_id = location.room_id

    if not target_room_id:
        return {"chatting_agents": []}

    # Get TRPG orchestrator to check chatting agents
    trpg_orchestrator = get_trpg_orchestrator()
    chatting_agent_ids = trpg_orchestrator.get_chatting_agents(target_room_id, agent_manager)

    # Get current streaming state
    streaming_state = await agent_manager.get_streaming_state_for_room(target_room_id)

    # Get agent details (filter out hidden agents like Action Manager)
    chatting_agents = []
    if chatting_agent_ids:
        all_agents = await crud.get_agents_cached(db, target_room_id)
        agent_map = {agent.id: agent for agent in all_agents}

        for agent_id in chatting_agent_ids:
            if agent_id in agent_map:
                agent = agent_map[agent_id]
                agent_state = streaming_state.get(agent_id, {})
                # Send actual agent name (frontend handles display logic)
                agent_info = {
                    "id": agent.id,
                    "name": agent.name,
                    "profile_pic": agent.profile_pic if not is_action_manager(agent.name) else None,
                    "thinking_text": agent_state.get("thinking_text", ""),
                    "response_text": agent_state.get("response_text", ""),
                }
                # For Action_Manager, include has_narrated flag so frontend can unblock input
                if is_action_manager(agent.name):
                    agent_info["has_narrated"] = trpg_orchestrator.has_narration_produced(target_room_id)
                chatting_agents.append(agent_info)

    # Check if World Seed Generator is active (during onboarding complete tool)
    seed_status = trpg_orchestrator.get_seed_generation_status(target_room_id)
    if seed_status:
        chatting_agents.append(
            {
                "id": -1,  # Virtual ID for World Seed Generator
                "name": seed_status.get("name", "World Seed Generator"),
                "profile_pic": None,
                "thinking_text": seed_status.get("thinking_text", "Creating your world..."),
                "response_text": seed_status.get("response_text", ""),
            }
        )

    # Check if a sub-agent (Summarizer, etc.) is active
    sub_agent_status = trpg_orchestrator.get_sub_agent_status(target_room_id)
    if sub_agent_status:
        chatting_agents.append(
            {
                "id": -2,  # Virtual ID for sub-agents
                "name": sub_agent_status.get("name", "Processing"),
                "profile_pic": None,
                "thinking_text": sub_agent_status.get("thinking_text", "Processing..."),
                "response_text": sub_agent_status.get("response_text", ""),
            }
        )

    return {"chatting_agents": chatting_agents}

File: backend/src/http/routes/game/polling.ts
/**
 * `GET /worlds/{id}/poll` — the endpoint the whole frontend lives on, running
 * every two seconds per open world.
 *
 * It is not a read. Because the agent side of the app writes to the
 * *filesystem*, this request also syncs the phase off `world.json` (what makes
 * the "Enter World" button appear), imports the seed generator's `player.json`,
 * writes the "<player> arrives at <place>" line and replays it to the Action
 * Manager as the world's first turn. All but the phase sync are one-shot: they
 * run only while the phase is `active` and the row has no
 * `current_location_id`, which the import fills in.
 */

import { Hono } from 'hono'
import { z } from 'zod'

import { getAgentsCached } from '@/crud/cached'
import { getLocation } from '@/crud/locations'
import {
  countAssistantMessages,
  createMessage,
  getMessages,
  getMessagesExcludingChat,
  getMessagesSince,
  type MessageWithAgent,
} from '@/crud/messages'
import { getPlayerState } from '@/crud/player-state'
import { getWorld, updateWorld } from '@/crud/worlds'
import type { Location, World, WorldPhase } from '@/db/schema'
import { isActionManager } from '@/domain/agent'
import { toLangKey } from '@/domain/enums'
import { getArrivalMessage } from '@/domain/localization'
import { getLogger } from '@/infrastructure/logging/logger'
import { parseJsonColumn } from '@/schemas/common'
import { PersistenceManager } from '@/services/persistence-manager'
import type { AppState } from '@/http/state'
import type { AppEnv } from '@/http/types'
import {
  boolQueryParam,
  deferBackground,
  intPathParam,
  intQueryParam,
  requireWorld,
  toPollMessage,
  worldsDirOf,
  type PollMessage,
} from './shared'

const logger = getLogger('GameRouter.Polling')

/** The `state` block of the poll response, read field-for-field by `usePolling`. */
interface PollState {
  stats: Record<string, unknown>
  inventory_count: number
  turn_count: number
  phase: WorldPhase
  /** Set by the onboarding `complete` tool; drives the "Enter World" button. */
  pending_phase: string | null
  is_chat_mode: boolean
  /** Where the transcript resumes after chat mode; see `./chat-mode`. */
  chat_mode_start_message_id: number | null
  game_time: { hour: number; minute: number; day: number } | null
}

interface PollResponseBody {
  messages: PollMessage[]
  state: PollState | null
  location?: { id: number; name: string }
  suggestions?: string[]
}

export function createPollingRoutes(state: AppState): Hono<AppEnv> {
  const routes = new Hono<AppEnv>()

  routes.get('/worlds/:world_id/poll', (c) => {
    const worldId = intPathParam(c, 'world_id')
    const sinceMessageId = intQueryParam(c, 'since_message_id')
    // Keeps the client on the onboarding room after the phase flips to
    // `active`, so the interview's last lines do not vanish.
    const pollOnboarding = boolQueryParam(c, 'poll_onboarding', false)

    let world = requireWorld(state, c, worldId)

    // A near-copy of `worlds.ts::syncWorldFromFs`, separate because the new
    // phase is needed *before* the write, to route this very request.
    const fsConfig = state.services.worlds.loadWorldConfig(world.name)
    let currentPhase: WorldPhase = world.phase ?? 'onboarding'

    if (fsConfig) {
      const updates: Parameters<typeof updateWorld>[2] = {}
      let changed = false

      if (fsConfig.phase !== world.phase) {
        logger.info(
          `Phase mismatch for world ${world.name}: DB=${String(world.phase)}, FS=${fsConfig.phase}. Syncing.`,
        )
        currentPhase = asWorldPhase(fsConfig.phase)
        updates.phase = currentPhase
        changed = true
      }
      if (fsConfig.userName && fsConfig.userName !== world.userName) {
        logger.info(
          `user_name mismatch for world ${world.name}: DB=${String(world.userName)}, FS=${fsConfig.userName}. Syncing.`,
        )
        updates.userName = fsConfig.userName
        changed = true
      }
      if (fsConfig.genre && fsConfig.genre !== world.genre) {
        logger.info(
          `genre mismatch for world ${world.name}: DB=${String(world.genre)}, FS=${fsConfig.genre}. Syncing.`,
        )
        updates.genre = fsConfig.genre
        changed = true
      }
      if (fsConfig.theme && fsConfig.theme !== world.theme) {
        logger.info(
          `theme mismatch for world ${world.name}: DB=${String(world.theme)}, FS=${fsConfig.theme}. Syncing.`,
        )
        updates.theme = fsConfig.theme
        changed = true
      }

      if (changed) {
        updateWorld(state.db, worldId, updates)
        world = getWorld(state.db, worldId)!
      }
    }

    let playerState = getPlayerState(state.db, worldId)

    let targetRoomId: number | null = null
    let location: Location | null = null

    if ((pollOnboarding || currentPhase === 'onboarding') && world.onboardingRoomId) {
      targetRoomId = world.onboardingRoomId
    } else if (currentPhase === 'active' && !pollOnboarding) {
      // A null `current_location_id` on an already-active world means the seed
      // generator finished and nothing has adopted its output yet.
      if (playerState && !playerState.currentLocationId) {
        const fsState = state.services.players.loadPlayerState(world.name)
        if (fsState?.currentLocation) {
          logger.info(`Syncing player state from filesystem for world '${world.name}'`)
          new PersistenceManager(
            state.db,
            worldId,
            world.name,
            worldsDirOf(state),
          ).syncPlayerStateFromFilesystem()
          playerState = getPlayerState(state.db, worldId)

          if (playerState?.currentLocationId) {
            const arrivalLocation = getLocation(state.db, playerState.currentLocationId)
            if (arrivalLocation?.roomId) {
              const locationName = arrivalLocation.displayName || arrivalLocation.name
              const language = toLangKey(world.language)
              const userName = world.userName || (language === 'ko' ? '여행자' : 'The traveler')
              const arrivalContent = getArrivalMessage(userName, locationName, language)

              createMessage(state.db, arrivalLocation.roomId, {
                content: arrivalContent,
                role: 'user',
                participantType: 'system',
                participantName: 'System',
              })
              logger.info(`Sent arrival message for '${userName}' at '${locationName}'`)

              const sceneRoomId = arrivalLocation.roomId
              deferBackground(
                async () => {
                  const taskWorld = getWorld(state.db, worldId)
                  if (!taskWorld) return
                  await state.orchestrator.handlePlayerAction({
                    world: taskWorld,
                    roomId: sceneRoomId,
                    action: arrivalContent,
                  })
                },
                { name: `poll_initial_scene:world=${worldId}` },
              )
              logger.info('Polling: Triggered initial scene generation after phase transition')
            }
          }
        }
      }

      if (playerState?.currentLocationId) {
        const current = getLocation(state.db, playerState.currentLocationId)
        if (current?.roomId) {
          location = current
          targetRoomId = current.roomId
          recoverStalledOpening(state, world, current, current.roomId, playerState.turnCount ?? 0)
        }
      }
    }

    // No room means nothing to show yet. The client keeps polling; not an error.
    if (!targetRoomId) {
      return c.json<PollResponseBody>({ messages: [], state: null })
    }

    // `chat_session_id` partitions the shared room; which half this poll wants
    // depends on the player's mode, except in onboarding, which shows all.
    const isChatMode = playerState?.isChatMode ?? false

    let messages: MessageWithAgent[]
    if (sinceMessageId) {
      messages = getMessagesSince(state.db, targetRoomId, sinceMessageId, 50)
      if (!isChatMode && !pollOnboarding) {
        messages = messages.filter((m) => m.chatSessionId === null)
      }
    } else if (isChatMode || pollOnboarding) {
      messages = getMessages(state.db, targetRoomId)
    } else {
      messages = getMessagesExcludingChat(state.db, targetRoomId)
    }

    // System messages are the game talking to itself; the player sees answers.
    const visible = messages.filter((m) => m.participantType !== 'system')

    const fsState = state.services.players.loadPlayerState(world.name)
    const gameTime = fsState?.gameTime
      ? {
          hour: fsState.gameTime.hour ?? 8,
          minute: fsState.gameTime.minute ?? 0,
          day: fsState.gameTime.day ?? 1,
        }
      : null

    const response: PollResponseBody = {
      messages: visible.map(toPollMessage),
      state: {
        stats: parseJsonColumn(playerState?.stats ?? null, z.record(z.string(), z.unknown())) ?? {},
        inventory_count:
          parseJsonColumn(playerState?.inventory ?? null, z.unknown().array())?.length ?? 0,
        turn_count: playerState?.turnCount ?? 0,
        // The synced phase, not the row's — the row may have been read before
        // the write above landed.
        phase: currentPhase,
        pending_phase: fsConfig?.pendingPhase ?? null,
        is_chat_mode: playerState?.isChatMode ?? false,
        chat_mode_start_message_id: playerState?.chatModeStartMessageId ?? null,
        game_time: gameTime,
      },
    }

    // Absent during onboarding rather than null: the key itself is only added
    // when a location is in play, and the frontend distinguishes the two.
    if (location) {
      response.location = { id: location.id, name: location.displayName || location.name }
    }

    // Always sent, even unchanged. `suggest_options` can write these *after* the
    // narration message is persisted but before the next poll, so a client that
    // fetched them only when messages arrived would miss the ones belonging to
    // the turn it just rendered.
    response.suggestions = state.services.rooms.loadSuggestions(world.name)

    return c.json(response)
  })

  /**
   * Typing indicators: real agents with a live session, plus two virtual ids
   * owning no room agent — the World Seed Generator (`-1`) and a Task sub-agent
   * (`-2`), which must differ because the frontend keys rows by id.
   *
   * **Gap: `thinking_text` and `response_text` are always empty.** They need a
   * per-room registry of partially-streamed responses; the SDK layer keeps that
   * on the turn instead. `has_narrated`, which unblocks input, is wired.
   */
  routes.get('/worlds/:world_id/chatting-agents', (c) => {
    const worldId = intPathParam(c, 'world_id')
    const pollOnboarding = boolQueryParam(c, 'poll_onboarding', false)
    const world = requireWorld(state, c, worldId)

    const targetRoomId = chattingAgentsRoom(state, world, pollOnboarding)
    if (!targetRoomId) return c.json({ chatting_agents: [] })

    const chattingAgentIds = state.orchestrator.getChattingAgents(targetRoomId)
    const chattingAgents: Record<string, unknown>[] = []

    if (chattingAgentIds.length > 0) {
      const byId = new Map(getAgentsCached(state.db, targetRoomId).map((a) => [a.id, a]))

      for (const agentId of chattingAgentIds) {
        const agent = byId.get(agentId)
        if (!agent) continue

        const info: Record<string, unknown> = {
          id: agent.id,
          name: agent.name,
          // The Action Manager narrates; an avatar would put a face on the prose.
          profile_pic: isActionManager(agent.name) ? null : agent.profilePic,
          thinking_text: '',
          response_text: '',
        }
        if (isActionManager(agent.name)) {
          info.has_narrated = state.orchestrator.hasNarrationProduced(targetRoomId)
        }
        chattingAgents.push(info)
      }
    }

    const seedStatus = state.orchestrator.getSeedGenerationStatus(targetRoomId)
    if (seedStatus) {
      chattingAgents.push({
        id: -1,
        name: seedStatus.name || 'World Seed Generator',
        profile_pic: null,
        thinking_text: seedStatus.thinkingText || 'Creating your world...',
        response_text: seedStatus.responseText,
      })
    }

    const subAgentStatus = state.orchestrator.getSubAgentStatus(targetRoomId)
    if (subAgentStatus) {
      chattingAgents.push({
        id: -2,
        name: subAgentStatus.name || 'Processing',
        profile_pic: null,
        thinking_text: subAgentStatus.thinkingText || 'Processing...',
        response_text: subAgentStatus.responseText,
      })
    }

    return c.json({ chatting_agents: chattingAgents })
  })

  return routes
}

/**
 * How stale an unanswered arrival line has to be before the opening counts as
 * dead rather than slow. Comfortably past the ~40s an opening turn takes, and
 * it is what makes the check safe against a turn that has been handed to
 * `deferBackground` but has not registered with the orchestrator yet — that
 * gap is a macrotask, this is two minutes.
 */
const OPENING_STALL_MS = 120_000

/**
 * Restart an opening scene that died before it narrated.
 *
 * `POST /worlds/{id}/enter` writes the arrival line and hands the first turn to
 * a background task with nothing behind it: a backend that stops in the ~40s
 * before the `narration` tool call — a watch reload, a Ctrl+C, a thrown turn —
 * leaves the room holding only that arrival message, which this endpoint
 * filters out as a system line. The player is then looking at an empty world
 * with no way back but a reset.
 *
 * Cheap checks first: this runs on every poll of every active world, and the
 * message count query is reached only by a world that has taken no turn and has
 * nothing running.
 */
function recoverStalledOpening(
  state: AppState,
  world: World,
  location: Location,
  roomId: number,
  turnCount: number,
): void {
  if (turnCount !== 0) return
  if (state.orchestrator.isBusy(roomId)) return
  if (countAssistantMessages(state.db, roomId) > 0) return

  // No arrival line means no turn was ever started for this room — a world mid
  // seed generation, not a stalled opening. Starting one here would narrate an
  // arrival the game has not staged yet.
  const arrival = getMessages(state.db, roomId).find(
    (message) => message.participantType === 'system',
  )
  if (!arrival) return

  const age = arrival.timestamp ? Date.now() - arrival.timestamp.getTime() : 0
  if (age < OPENING_STALL_MS) return

  // One restart per room per process: a turn that keeps ending without
  // narrating must not be relaunched by every poll two seconds later. A reset
  // mints a new room, so a deliberate retry always gets a fresh claim.
  if (!state.orchestrator.claimOpeningRestart(roomId)) return

  logger.warning(
    `Opening scene for '${world.name}' never produced narration (room ${roomId}); restarting it`,
  )

  deferBackground(
    async () => {
      const taskWorld = getWorld(state.db, world.id)
      if (!taskWorld) return
      await state.orchestrator.handlePlayerAction({
        world: taskWorld,
        roomId,
        action: arrival.content,
      })
    },
    { name: `recover_opening:world=${world.id}` },
  )
  logger.info(`Polling: Restarted the opening scene at '${location.name}'`)
}

/**
 * Deliberately *not* `/poll`'s room selection: the row's phase, not the
 * filesystem's, and no player-state sync. A phase flipped on disk but not in the
 * DB costs one cycle of indicators against the old room.
 */
function chattingAgentsRoom(state: AppState, world: World, pollOnboarding: boolean): number | null {
  if ((pollOnboarding || world.phase === 'onboarding') && world.onboardingRoomId) {
    return world.onboardingRoomId
  }
  if (pollOnboarding) return null

  const playerState = getPlayerState(state.db, world.id)
  if (!playerState?.currentLocationId) return null
  return getLocation(state.db, playerState.currentLocationId)?.roomId ?? null
}

/** `world.json` is hand-editable; anything unrecognised reads as onboarding. */
function asWorldPhase(phase: string): WorldPhase {
  return phase === 'active' || phase === 'ended' ? phase : 'onboarding'
}
